mcstatus/motd.py

```diff
@@ -0,0 +1,558 @@
+
+    @property
+    def xterm_256(self) -> str:
+        """Getter for a xterm-256 variant of MOTD.
+
+        The only difference from ``ansi`` is that it's using RGB colors instead of ``red``, ``blue`` etc.
+
+        :returns: String full of xterm-256 colors.
+        """
+        result = r"\e[0m"
+        for element in self.parsed:
+            if isinstance(element, str):
+                result += element
+            elif isinstance(element, TranslationTag):
+                continue
+            elif isinstance(element, WebColor):
+                result += r"\x1b[38;2;{0};{1};{2}m".format(*element.rgb)
+
+            elif isinstance(element, Formatting):
+                if element == Formatting.RESET:
+                    result += r"\e[0m"
+                else:
+                    tag_number = FORMATTING_TO_ANSI_TAGS[element]
+                    result += r"\e[" + tag_number + "m"
+            else:
+                mc_color_to_hex = MINECRAFT_COLOR_TO_HEX_BEDROCK if self.bedrock else MINECRAFT_COLOR_TO_HEX_JAVA
+                foreground_color, background_color = mc_color_to_hex[element]
+
+                result += r"\x1b[38;2;{0};{1};{2}m".format(
+                    *tuple(int(foreground_color.lstrip("#")[i : i + 2], 16) for i in (0, 2, 4))
+                )
+                result += r"\x1b[48;2;{0};{1};{2}m".format(
+                    *tuple(int(background_color.lstrip("#")[i : i + 2], 16) for i in (0, 2, 4))
+                )
+
+        return result + r"\e[0m"
+
+    @property
+    def ansi(self) -> str:
+        """Getter for an ANSI colors variant of MOTD.
+
+        The only difference from ``xterm_256`` is that it's using ``red``, ``blue`` etc. instead of RGB colors.
+        This also using ``xterm_256`` colors, if found ``WebColor`` in ``Motd.parsed``.
+
+        .. note:: ``MINECOIN_GOLD`` will be replaced by ``GOLD`` here.
+
+        .. note:: This don't check non ANSI characters, only colors.
+
+        .. warn:: There is no support for background colors, use ``xterm_256`` instead.
+
+        :returns: String full of ANSI colors.
+        """
+        result = r"\e[0m"
+        for element in self.parsed:
+            if isinstance(element, str):
+                result += element
+            elif isinstance(element, TranslationTag):
+                continue
+            elif isinstance(element, WebColor):
+                result += r"\x1b[38;2;{0};{1};{2}m".format(*element.rgb)
+
+            elif isinstance(element, Formatting):
+                if element == Formatting.RESET:
+                    result += r"\e[0m"
+                else:
+                    tag_number = FORMATTING_TO_ANSI_TAGS[element]
+                    result += r"\e[" + tag_number + "m"
+            else:
+                foreground_color = MINECRAFT_COLOR_TO_ANSI[element]
+
+                result += r"\e[0;" + foreground_color + "m"
+
+        return result + r"\e[0m"
+
+
+class MotdSimplifies:
```

Comment:
I don't really like this. The individual simplify methods are popping from a list that we're iterating over, that's never a good idea. 

Even if we're iterating over a copy here, that might actually cause even more issues, as the `enumeration` that these functions receive is simply the index number on that copy, however the moment we pop out any value, all of the indices change. 

Say we're at index 5 that's popped, now we go to index 6 in a copy, but in the original, where we popped from, index 5 was removed, and now became index 6, which means index 6 is now actually index 7, etc. Its' just really messy.

Instead, I'd go for an appraoch that simply returns the indices to be removed, like this:
```python
    def _simplify_doubled_items(self) -> set[int]:
        """Get indices of all doubled items that can be removed."""
        to_remove = set()

        for index, item in enumerate(self.parsed):
            try:
                next_item = self.parsed[index + 1]
            except IndexError:  # Last item (without any next item)
                break

            if item == next_item:
                to_remove.add(index)

        return to_remove
        
    def _simplify_doubled_colors(self) -> set[int]:
        """Get indices of all doubled color items.

        As colors (obviously) override each other, we only ever care about the last one, ignore
        the previous ones. (for example: specifying red color, then orange, then yellow, then some text
        will just result in yellow text)
        """
        to_remove = set()

        prev_color = None
        for index, item in enumerate(self.parsed):
            if isinstance(item, (MinecraftColor, WebColor)):
                # If we found a color after another, remove the previous color
                if prev_color is not None:
                    to_remove.add(prev_color)
                prev_color = index

            # If we find a string, that's what our color we found previously applies to,
            # set prev_color to None, marking this color as used
            if isinstance(item, str):
                prev_color = None

        return to_remove
```

Which can then simply be unioned together like
```python
def simplify(self) -> Self:
    to_remove = set()
    to_remove.union(self._simplify_doubled_colors())
    to_remove.union(self._simplify_doubled_items())
    
    parsed = self.parsed.copy()

    # To safely remove multiple indices by popping, we need to go from last index
    # in a sorted order, so that the indices don't change on us.
    for index in sorted(to_remove, reverse=True):
        parsed.pop(index)
    
    return __class__(parsed)
```

Notice that in this example, the individual simpify methods actually belong to `Motd` class itself. While this does introduce some clutter in that class, it's much less weird than having a class like `MotdSimplifiers` that's basically just a namespace for a bunch of simplification methods.

Reply:
> The individual simplify methods are popping from a list that we're iterating over, that's never a good idea.

Why not? It's much faster because iterating in Python is really slow. Like, it's faster in ~5 times because every simplifier is just if-else statements that gives no effect on speed, while new iteration over `self.parsed` takes the same time, as current `Motd.simplified()` call.

Yes, it would not be so important, if count that MOTDs are limited to ~500 symbols but 5 times difference is still huge.

> Even if we're iterating over a copy here, that might actually cause even more issues, as the `enumeration` that these functions receive is simply the index number on that copy, however the moment we pop out any value, all of the indices change.

Hooks don't have access to original `parsed` list, and **must not** save any info about previous elements. They must use `parsed[enumeration - 1]` to not break anything.

And the only method, that has access to original `parsed` list is `Motd.simplified`, which also must not depend on enumeration.

> Say we're at index 5 that's popped, now we go to index 6 in a copy, but in the original, where we popped from, index 5 was removed, and now became index 6, which means index 6 is now actually index 7, etc. Its' just really messy.

It's impossible if we will follow rules that I mentioned upper.

> it's much less weird than having a class like `MotdSimplifiers` that's basically just a namespace for a bunch of simplification methods.

This class exists for possible extending by users with monkeypatching.

Reply:
I'm aware that these hook methods don't modify the original `self.parsed`, I wouldn't have approved this otherwise, but the issue here isn't the original getting changed, the issue is that we might be popping wrong elements from the copied list.

### Skipping elements problem

Let me try to demonstrate this with an actual example in python:
![image](https://user-images.githubusercontent.com/20902250/206928415-6574291d-c05e-42c4-84d6-080db392aef0.png)

Notice that, in this example, since we popped out an element at index 2, the list was shrunk and element that was before on index 3 suddenly moved to index 2, to replace the popped element (as there can't be a gap in lists indices). This meant that as we iterated over the next element, the code thinks we've already got past index 2, and now gets the element at index 3, which is the number 5. But wait, we've now skipped number 4!

Yes, this is covered by the while loop above, re-running the code, so in the end we will go through all elements. That said though, this is still some very spooky logic of iterators that we're relying on here, and even though it actually somehow works, it doesn't mean that it's a good idea.

### Another problem

Yet another problem that your approach has is that you run the hooks in a loop, without checking whether the hook has already removed that item. This means that one hook could end up removing the item you're at (say item at index 3), and another hook, that also evaluates this item as irrelevant will make yet another deletion of an item at index 3. This time however, as the original item was already deleted, and the item after took it's place, we've just removed the next item, that was never supposed to get removed in the first place.

Yes, this could be safe-guarded by having a bool flag indicating whether something was removed, and skipping the rest of the hooks for that element, however that then causes the while loop to repeat much more times, increasing the time this takes. Not to mention that the first problem I've pointed out still remains.

### Runtime

The speed benefit doesn't really matter almost at all here, we're dealing with really small amount of elements, and the `simplified` method will almost certainly only be called once, if at all by the users. But anyway, let's take a look then.

Since the while loop in your example means that we're going to be iterating over the entire list of items more than once anyway, it will give us `O(x*n)` runtime, where `x` is the number of times we need to rerun until no further modifications are required. With my approach, we also get runtime of `O(x*n)`, but here, `x` is the number of simplification functions, as each function can run in `O(N)` time, and we run `x` such functions. 

This makes my approach scale similarly, and in some cases, it might even perform better as again, the functions only run once and collect all of the elements that were found as irrelevant and should be excluded, whereas with your approach, if there's too many cases where the rule keeps removing 1 element only, like say with: `Formatting, Formatting, Formatting, Color, hitting the doubled_colors hook 3 times to remove all 3 formattings before colors, requiring 3 iterations from just this one hook.

This alone might still seem like the approach you're using might be better, or at least, equal, but in actuality, your approach is much worse, due to the element popping on n-th index. This kind of pop means the list will need to move all of the remaining elements to the left. This gives us a worst case of another `O(N)` for this moving operation in these lists. Additionally, as python uses dynamically sized lists, the list might even get shrunk during this removal, which would mean moving all elements into another memory location, which has smaller allocation, and destroying the original allocation. This shrinking is always `O(N)`.

However with my approach, all that we're doing is collecting the indices that need to be removed, and while in my example I just sorted these indices and removed them one by one, a much faster approach would be to just loop over the list, and start adding all items into some other result list, skipping the indices we've obtained. This is then simply another single iteration over the motd items.

### Extensibility

Additionally, extending by monkey patching is not something the users should have to do, in a well written OOP version, all the user should have to do is override a function/modify a variable. In this case, in my example, I've shown how we could just call the functions manually and union them with the `to_delete` set, however we could just as well have a class variable holding all of these functions, or something similar, to provide this extensibility.

Reply:
I agree, thanks for explanation. Looks like, when I wrote previous answer, I forgot the code, and forgot that my solution iterates list few times as well.

I will try to implement your solution as well as test that will find this bug.

And for extensibility, I would better like to move methods/properties as `simplified`, `ansi` and `xterm_256` to its own class, and then get this class from variable.
```py
class Motd:
    SIMPLIFY_CLASS = Simplify
    ANSI_TRANSFORMER_CLASS = AnsiTransformer
    ...
    @property
    @cached
    def _simplify(self) -> Simplify:
        return self.SIMPLIFY_CLASS(...)

    @property
    @cached
    def _ansi_transformer(self) -> AnsiTransformer:
        return self.ANSI_TRANSFORMER_CLASS(...)
```
So then user can subclass `Simplify`/`AnsiTransformer` class and replace it with
```py
status = server.status()
status.motd.SIMPLIFY_CLASS = MySimplify
motd = status.motd.simplify()
```

Reply:
Hooks with sets were implemented [here](https://github.com/py-mine/mcstatus/compare/8a4b71e2c9c2e3b8fc8cd8f6cf92a6340095795b..6f42759c97409b8fefb379a76077bbb32b2a508c).

Reply:
I do understand the desire to split these up, as the motd class is pretty big with all of this, but I'm honestly not sure if it's really such a good idea. It's pretty odd to see this kind of splitting that relies on caching instances with properties and it's just not really pythonic.

If at all, I'd rather see mixin classes do this then using this hacky composition scheme. That could look like:
```python
@dataclass
class Motd(_MotdSimplifiesMixin, _MotdTransformersMixin):
    ...  # Includes constructors

# This is NOT a dataclass    
class _MotdSimplifiesMixin:
    __slots__ = tuple()
    
     # type definition for `parsed` (as this is a mixin to Motd, which we know does contain this, so it will be available)
    parsed: list[MotdElement]
    
    ...  # This should include the `simplified` method, along with all of the individual simplifies


# Same approach for transformers class
```

Mixins are still not the greatest solution, but they're much more pythonic than composition like you suggested, and while they're still result in the Motd class being pretty big, at least in code, it's split up so it's maintainable and more readable.

Reply:
I don't like the idea of any such `Enum` class, that's even more unpythonic than your original suggestion (composition).

I wouldn't mind making `motd` a package though, but why? It could separate the color classes and the formatting, however it wouldn't help at all with splitting up the motd class itself, that would still have to include all of the transformations and simplification methods.

Reply:
> I wouldn't mind making `motd` a package though, but why? It could separate the color classes and the formatting, however it wouldn't help at all with splitting up the motd class itself, that would still have to include all of the transformations and simplification methods.

Without mixins, it will. Like `Motd` will have only proxies, [as it does now](https://github.com/py-mine/mcstatus/pull/335/files#diff-c462c16c38f067b712f30e9e5d65eda0a525e97348b12492eacf676da5656ed6R405).

Reply:
I found that your solution doesn't work well, look at example
```py
        obj = Motd([Formatting.BOLD, "", Formatting.RESET, "", MinecraftColor.RED, ""])
>       assert obj.simplified() == Motd([])
E       AssertionError: assert Motd(parsed=[<Formatting.BOLD: 'l'>, <Formatting.RESET: 'r'>, <MinecraftColor.RED: 'c'>], bedrock=False) == Motd(parsed=[], bedrock=False)
```
So in this example, to actually simplify the motd, you need to call `Motd.simplified()` twice.

Reply:
Hmm, I don't think there's any good way to go around this. We could prioritize the rules, to minimize this issue, but it might not be possible to entirely avoid these situations. 

For that reason, it might be a good idea to bring back the while loop that repeats the simplifications until the motd elements stop changing. Obviously, this decreases the efficiency a lot, but it's not really that relevant due to this method likely only ever being called once on any motd, and it's still much better than the original approach.

Reply:
So I undo all changes with sets and IDs to delete, and bring back old logic? It was much simpler for this kind of architecture.

Reply:
No, as I said, the original logic was incredibly flawed as it was manipulating the lists while iterating over them, and the index based removals were costly performance wise, making it O(M*N^2) (each iteration, we might remove element at arbitrary index, which is O(N)), that all was happening M times (until motd stops changing). 

The new approach is O(M*N), we repeat M times, and the individual processes each just take O(N), the final removal then takes another O(N), but O(N) + O(N) + O(N) + ... is still just O(N), not O(N^2), making the new approach still win over the previous one in performance. On top of that, it doesn't introduce any hacky removals during iteration, which is an anti-pattern which should be avoided in pretty much ALL cases.

Reply:
Fixed [here](https://github.com/py-mine/mcstatus/compare/5a99136ba6ccfe68f2ad9ac32efeb5742a8ef097..33969fa91ce672da8406b30686151ffb9113ab62) but do not resolve it for now. I will add test (maybe tomorow).

Reply:
Added test [here](https://github.com/py-mine/mcstatus/compare/33969fa91ce672da8406b30686151ffb9113ab62..c5d0a9f4db39c4962022ae57ab6e295f1faceab0).